### PR TITLE
refactor(api): rewrite api structure, make it more clean

### LINKED ISSUE #148: test: create tests for packages/api/lib/db.js
@sayak-sarkar would start working on it as he is done with API keys.

### LINKED ISSUE #224: Occasional api test failure, "bad file descriptor, write"
This happens intermittently on my machine, logging for future reference.

```
10:40:51 {nested-config} ~/projects/spaship/packages/api
$ npm test

> @spaship/api@0.5.0 test /home/mclayton/projects/spaship/packages/api
> ../../node_modules/.bin/jest --restoreMocks

 PASS  lib/db.apikey.test.js
  api.db.apikey
    attach
      ✓ should be able to attach to the collection (8ms)
    apikey CRUD
      ✓ should be able to create a new api key (8ms)
      ✓ should be able to get api keys by userid (6ms)
      ✓ should be able to get userid by apikey (7ms)
      ✓ should be able to delete a key (5ms)

 FAIL  lib/background/autosync.test.js
  api.autosync
    syncTarget
      lifecycle
        ✕ should start and stop (7ms)
        ✓ should sync targets while running (1ms)
        ✓ should sync targets periodically per target's timer
      disk cache
        ✓ should save to disk the response from a sync target (4ms)
        ✓ should save to disk the response from a sync target with subpaths (9ms)

  ● api.autosync › syncTarget › lifecycle › should start and stop

    EBADF: bad file descriptor, write



      at actualWrite (../common/node_modules/sonic-boom/index.js:272:24)
      at SonicBoom.Object.<anonymous>.SonicBoom.write (../common/node_modules/sonic-boom/index.js:154:5)
      at Object.write (../common/node_modules/pino/lib/tools.js:240:12)
      at Pino.write (../common/node_modules/pino/lib/proto.js:161:15)
      at Pino.LOG [as info] (../common/node_modules/pino/lib/tools.js:39:26)
      at Autosync.start (lib/background/autosync.js:1520:9)
      at Object.<anonymous> (lib/background/autosync.test.js:69:12)

 PASS  lib/metadata.test.js
  api.metadata
    write
      ✓ should add metadata to a deployed SPA (27ms)
    getAll
      ✓ should retrieve metadata for all deployed SPAs (6ms)
      ✓ should return an empty array if the webroot can't be accessed (1ms)
    get
      ✓ should retrieve metadata when given a spa's directory (1ms)

--------------------|----------|----------|----------|----------|-------------------|
File                |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
--------------------|----------|----------|----------|----------|-------------------|
All files           |     74.5 |    51.52 |    65.63 |    75.86 |                   |
 api                |    71.43 |       50 |     37.5 |    70.37 |                   |
  config.js         |    71.43 |       50 |     37.5 |    70.37 |... ,96,98,102,104 |
 api/lib            |    94.34 |    61.54 |      100 |    94.34 |                   |
  db.apikey.js      |      100 |    66.67 |      100 |      100 |                14 |
  db.js             |       75 |       50 |      100 |       75 |                10 |
  db.mock-mongo.js  |    88.89 |     62.5 |      100 |    88.89 |             12,22 |
  metadata.js       |      100 |      100 |      100 |      100 |                   |
 api/lib/background |    60.29 |    43.75 |    53.85 |    63.08 |                   |
  autosync.js       |    60.29 |    43.75 |    53.85 |    63.08 |... 11,112,131,141 |
--------------------|----------|----------|----------|----------|-------------------|
Test Suites: 1 failed, 2 passed, 3 total
Tests:       1 failed, 13 passed, 14 total
Snapshots:   0 total
Time:        1.466s, estimated 2s
Ran all test suites.
{"message":"ENOENT, no such file or directory '/fake/webroot'","code":"ENOENT","errno":-2,"path":"/fake/webroot"}
npm ERR! Test failed.  See above for more details.
```

### LINKED ISSUE #260: convert all API endpoints to JSend
https://github.com/omniti-labs/jsend

Nice, very simple standard response format for JSON APIs.

### LINKED ISSUE #264: (api) should support api versioning
Our recent target is MVP, but there should be have a lot of changes soon or later.

API versioning  could make the end user with different version of cli could complete their request to SPAship api

### PR DESCRIPTION
**Pull request by kunyan**
* Remove autosync #54
* Use [mongoose](https://mongoosejs.com/) to replace native mongodb client
* Rewrite all middlewares
* Add error handler
* Fix #264  support API versioning
* Fix #260 convert all API endpoints to JSend
* Fix #224 remove log in fileService to make the UT work

I think also resolve #148 . we can do controller unit testing now

### The new API structure:

* `/api/v1/applications`
  * GET `/` list all applications
  * GET `/{applicationName}` get application info
  * POST `/deploy` add a new application and upload to deploy
  * POST `/` add new application
* `/api/v1/apiKeys`
  * GET `/` get all apiKeys of current login user
  * POST `/` add new apiKey
  * DELETE `/{label}` remove a apiKey by label

### The project folder structure
* controllers
  * all controllers in this directory
* middlewares
  * authorization
  * responseWrapper
  * errorHandler
* models
  * apiKey
  * application
* services
* routes
* utils

### CODE REVIEW

**Comment by kunyan** [comment]
I'll deploy this PR to our Dev environment

**Comment by kunyan** [comment]
@Jared-Sprague 
1. Can we do it post MVP ?
I was plan to do it in post MVP. but some changes if we do it later might be very difficult to turn around. Like api versioning, response format, authorization header.  In order to protect every logic is same like what we did before the refactor, I deployed this PR to Dev environment and asked @tanyutu to do the API testing.
2. AutoSync removed. 
sorry I didn't describe the replacement in description. I'm not just remove it at all. I have a similar project with is do the same thing, I'm plan to push it in mono repo after this PR merged. it has been tested in Labs production environment, it could be trusted.

**Comment by mwcz** [comment]
Hi @kunyan, since this is an open source project, here I am. :)

As I'm not officially working, I don't have time to do a thorough review, but I did look at every line of changes and here are a few cursory notes:

 - you seem to have deleted and recreated most of the existing files, rather than moving them (with `git mv`).  That severs the files' git history and we can never look at previous commits to that file again.  Please fix this.
 - don't forget that the create API key endpoint should not accept API key auth.  You requested this yourself and I agree.  I had that built in but it seems you removed it by applying the same auth rules to every endpoint.


There are a lot of really great changes in this PR, but my two cents is that since the MVP deadline has already passed (weeks ago), it's unwise to do a huge overhaul.  MVP isn't 1.0, and changes to the API are fair game during MVP.  But I'm on PPL and I'll leave it to the team to decide how to handle MVP.

**Comment by kunyan** [comment]
@mwcz Thanks so much to review this changes !

* I recreate most of files and keep the exists files of them when I developing. then removed the old code before push commit. sorry for not notice it will break git commit history :(  let me try if I can fix it
* Yes, the API key can not create/delete API key. this part is in my workspace now, but not push yet.

For MVP target. You will see only 3 important APIs.
1. deploy application
    This is almost not change...
2. list application
   This is almost not change... 
3. apiKey
   It is the most different codes here. so I added fully controller level unit test here to make sure everything works


I was plan to push this change earlier, there has some other urgent work blocks me :(.
I'm totally agree we should not have so big changes before the MVP since the target day is soon.
so I checked every details of this changes and deployed it to Dev. make sure the whole workflow could running as expect.

**Comment by sayak-sarkar** [comment]
@kunyan From a functionality stand-point, this looks good to me. Going through the previous comments, you have already answered most of the queries I had. I also agree with your reasoning about some points such as API versioning, which I also think should be rolled out before MVP in order to avoid having to make major changes to the overall structure right after MVP.

**Comment by Jared-Sprague** [comment]
I can't approve this until I see the replacement for AutoSync.    Why would you re-implement autosync?  Why not just use what I had already done and spent a lot of time one without even talking to me?

**Comment by kunyan** [comment]
@Jared-Sprague . Sorry for make you that feeling.
Actually, I was think #54 might be need take some time. and our MVP customer actually not using cp head/header/footer. So I planned use a exist cron job (I think I have introduce it when we start this project. it only without forceSync feature) to replace it. if someone want to test the SSI feature, we should make it work.

But you are right. I should not take the decision  by myself. and I reviewed the autoSync code, it is very easy to split out.
I'll push a changes for it. please wait for 1 hour.

**Comment by Jared-Sprague** [comment]
@kunyan Thanks!  Yeah I think cron would work, but it's not as flexible as AutoSync and configuring cron can vary between platforms.  I wrote AutoSync as a service that comes with spaship so people would not have to create cron jobs and be consistent way of syncing SSI files no matter what infrastructure you're running on.

Yeah if you split out AutoSync into it's own module, that can run in it's own pod, that would be perfect.  That would also solve the problem of having scaling the API pods, we can scale up the API pods without scaling the AutoSync processes.

**Comment by kunyan** [comment]
@Jared-Sprague autosync back and move to own packages. fix #54

**Comment by Jared-Sprague** [comment]
Ok we talked about this in our weekly call and Kun demo'd the API and the how AutoSync is working as it's own service now.  Like with any large PR, regression testing to make sure that all previously working functionality is still working is super important.  Kun says that Wei is working on this testing.  So once that is done and all test passing we can merge this.

File: packages/cli/src/commands/init.js [hunk 1]
```diff
@@ -27,44 +27,42 @@ class InitCommand extends Command {
...
+          when: showQuestions,
         },
         {
           name: "single",
           message: "Single page?",
           type: "confirm",
-          when: showQuestions
-        }
+          when: showQuestions,
+        },
```
[the remark this answers is not in the document]
Comment (kunyan) [reply packages/cli/src/commands/init.js:49]: @sayak-sarkar I think it caused by prettier upgraded. and I'm not the only one who has been impact. check this: https://github.com/spaship/spaship/pull/266/files#diff-88224a81a7099ed96801c7d8ae634ad6

**Comment by mwcz** [comment]
Dangling commas are good.  Dangling commas allow you to reorder elements without having to change the punctuation. Can you imagine how many fewer merge conflicts we would have if package.json allowed trailing commas? Besides, we use prettier and that alone should mean we never have to have this type of conversation. We aren't writing ES5 anymore.
-- 
Michael Clayton
https://clayto.com

**Comment by Jared-Sprague** [comment]
@mwcz that was exactly what I said on our call this morning,  we shouldnt be talking about stuff like this because prettier should take care of it for us.  So the conversation shifted to why prettier is not automatically keeping our formatting consistent like it should be.

**Comment by mwcz** [comment]
Great, yeah.  Prettier's defaults changed.  If prettier's defaults change again in the future, we can avoid this by not upgrading prettier anymore.
-- 
Michael Clayton
https://clayto.com

**Comment by tanyutu** [comment]
@kunyan 
API test is done for this version. please check the issues below:
1. use post /applications to create a new application entity returns 200 as successfully. But can't find this application record in get /applications. is that the right behavior?
2.just return text of "remove success" when delete /apiKey/{label}
3.delete /application/{name} returns 404
4. no "timestamp" show in body when post /applications
5. no "timestamp" show in body when post /applications/deploy
6. no "ref" show in body when post /applications

**Comment by kunyan** [comment]
> @kunyan
> API test is done for this version. please check the issues below:
> 
> 1. use post /applications to create a new application entity returns 200 as successfully. But can't find this application record in get /applications. is that the right behavior?


This is design for post MVP. user is able to create an application entity first, without upload any file.
But our GET /applications is for MVP only. it's return the data from disk, not from database

> 2.just return text of "remove success" when delete /apiKey/{label}


I fixed it in latest changes
the remove response will be
```json
{
  "status": "success",
  "message": "remove success"
}
```
> 3.delete /application/{name} returns 404

This has been fix in latest change.
> 4. no "timestamp" show in body when post /applications


This has been fix in latest change.
> 5. no "timestamp" show in body when post /applications/deploy


This has been fix in latest change.
> 4. no "ref" show in body when post /applications


When user create a new application entity, it won't need they input a specific branch or tag name. the `ref` will not store into database in that time, we should think about how to do version management in the future

**Comment by tanyutu** [comment]
I've retested, all issue have been fixed, except below one, please check.
when try to delete /application/{appname},always meet fail as can not found application by name, even the appname really existed in list.

**Comment by kunyan** [comment]
@mwcz the current JWT token will expired in 5mins, it makes @tanyutu very difficult to do API test
So we will not add the  request limit for APIKey access `/apiKeys*` for now

I filed a issue to make sure it will be acknowledge #300   and fixed before MVP release

File: packages/sync/index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+const AutoSync = require("./lib/autoSync");
+
+const port = config.get("port");
+
+const requestListener = function (req, res) {
+  if (req.url === "/force" && req.method === "GET") {
+    const autoSync = new AutoSync();
+    autoSync.forceSyncAll();
+    res.writeHead(200);
+    res.end("Forcing all autoSync targets to sync now");
```
[the remark this answers is not in the document]
Comment (kunyan) [reply packages/sync/index.js]: it has been fix

**Pull request merged**